controllers/ibu_controller.go

```diff
@@ -194,26 +204,40 @@ func (r *ImageBasedUpgradeReconciler) Reconcile(ctx context.Context, req ctrl.Re
 					return
 				}
 				if !isValid {
-					err = utils.UpdateIBUStatus(ctx, r.Client, ibu)
+					ibu.Status.ValidNextStages = getValidNextStageList(ibu, isAfterPivot)
+					if updateErr := utils.UpdateIBUStatus(ctx, r.Client, ibu); updateErr != nil {
+						r.Log.Error(updateErr, "failed to update IBU CR status")
+					}
 					return
 				}
 			}
-			nextReconcile = requeueImmediately()
-		}
-	} else {
-		inProgressStage := utils.GetInProgressStage(ibu)
-		if inProgressStage != "" {
-			nextReconcile, err = r.handleStage(ctx, ibu, inProgressStage)
-			if err != nil {
-				_ = utils.UpdateIBUStatus(ctx, r.Client, ibu)
-				return
+			ibu.Status.ValidNextStages = getValidNextStageList(ibu, isAfterPivot)
+			// Update status
+			if updateErr := utils.UpdateIBUStatus(ctx, r.Client, ibu); updateErr != nil {
+				r.Log.Error(updateErr, "failed to update IBU CR status")
 			}
+		}
+	}
+
+	inProgressStage := utils.GetInProgressStage(ibu)
+	if inProgressStage != "" {
+		nextReconcile, err = r.handleStage(ctx, ibu, inProgressStage)
+		if err != nil {
 			ibu.Status.ValidNextStages = getValidNextStageList(ibu, isAfterPivot)
+			if updateErr := utils.UpdateIBUStatus(ctx, r.Client, ibu); updateErr != nil {
+				r.Log.Error(updateErr, "failed to update IBU CR status")
+			}
+			return
 		}
 	}
 
+	ibu.Status.ValidNextStages = getValidNextStageList(ibu, isAfterPivot)
+
 	// Update status
-	err = utils.UpdateIBUStatus(ctx, r.Client, ibu)
+	if updateErr := utils.UpdateIBUStatus(ctx, r.Client, ibu); updateErr != nil {
```

Comment:
I am more concerned with this one, comparing to the update error in the middle. Are we sure that it's good idea to hide the error from the controller? Normally this is entitled for another reconcile.